### ISSUE TITLE
State sync from local snapshot: Implementation

### ISSUE THREAD

**Issue by sergio-mena** [cometbft/cometbft#29]
Tasks:

- [ ] Implementation (existing implementation of ADR083: tendermint/tendermint#9541)
- [ ] Introduce necessary changes to documentation and/or spec to keep them up to date
- [ ] Check if tendermint/tendermint#4642 can be closed once implementation merged

DoD:
- Implementation code complete and all PRs merged
- Any changes needed to the documentation/spec are merged

Original issue: tendermint/tendermint#9948

**Comment by chillyvee** [comment]
Happy to collaborate when you are ready.  We just updated the code for the latest Juno v13 and have some intermediate feedback based on previous discussions.

**Comment by adizere** [comment]
Thank @chillyvee, appreciate it ! We'll get back to you in a few weeks time after we have explored all the existing material in tendermint/tendermint and have clarity over the solution. This work is part of our Q2 priorities.

**Comment by adizere** [comment]
Hi @chillyvee would you kindly reconstruct your [original PR](https://github.com/tendermint/tendermint/pull/9541/commits/96202a30ad76ae31729abea2db84b3ddae24febf) and submit to this repo, with a base branch v0.34? We've now in the "understand problem & solution" space (#28) so it would be good to see this implementation.

Thanks in advance!

**Comment by yihuang** [comment]
> Hi @chillyvee would you kindly reconstruct your [original PR](https://github.com/tendermint/tendermint/pull/9541/commits/96202a30ad76ae31729abea2db84b3ddae24febf) and submit to this repo as a draft PR preferably, with a base branch v0.34? We've now in the "understand problem & solution" space (#28) so it would be good to see this implementation.
> 
> Thanks in advance!

Love this feature, would like to help if needed, BTW, where can I find the ADR083?

From my understanding, state-sync can be decoupled into different components:
- snapshot discovery and chunks downloading
- snapshot chunks verification with light client protocol
- snapshot restoration
- node startup with the new state.

I think ideally we should have offline commands for each steps to provide maximum flexibility.

**Comment by cason** [comment]
I think Adi referred to the open PR in the old repository: https://github.com/tendermint/tendermint/pull/9651

**Comment by yihuang** [comment]
> I think Adi referred to the open PR in the old repository: [tendermint/tendermint#9651](https://github.com/tendermint/tendermint/pull/9651)
> 
> This has become a huge PR due to the changes made in the upstream repository. We are only interested on this file: https://github.com/tendermint/tendermint/pull/9651/files#diff-393baba5ea65c4316e47d1a3715a1c4a8e934c1310071f982d2cf5418ef84d80

Thanks.

> Allow applications to start with a bootstrapped state alongside an empty Tendermint instance using a subsystem of the state sync protocol.

This is exactly what I want.

**Comment by yihuang** [comment]
Just did a squash rebase and a few tweaks to fix building.

**Comment by adizere** [comment]
Thank you @yihuang. I'll be opening an ADR based on the original one soon.

**Comment by adizere** [comment]
Ported the ADR here: https://github.com/cometbft/cometbft/pull/729. It's not yet clear to me if it's reflective of the solution, please have a look.

**Comment by chillyvee** [comment]
Seeing the activity here.  Need to review notes and will be back soon.

**Comment by yihuang** [comment]
> Seeing the activity here.  Need to review notes and will be back soon.

Please feel free to take over the PR when you are back 😄

**Comment by cason** [comment]
Any reason for targeting this change to branch `v0.34.x`? I would say it would be better and safer to target the `main` branch, then after due checks, backport it to release branches.

**Comment by chillyvee** [comment]
Is there a preference for v0.34 or main?  

Can probably make a working example for Juno v14 if anyone thinks that is helpful for the "understand" part of the work.

**Comment by yihuang** [comment]
> Any reason for targeting this change to branch `v0.34.x`? I would say it would be better and safer to target the `main` branch, then after due checks, backport it to release branches.

target main branch: https://github.com/cometbft/cometbft/pull/733

**Comment by adizere** [comment]
> Is there a preference for v0.34 or main?
> 
> Can probably make a working example for Juno v14 if anyone thinks that is helpful for the "understand" part of the work.

Excellent idea!

**Comment by adizere** [comment]
> Can probably make a working example for Juno v14 if anyone thinks that is helpful for the "understand" part of the work.

Any progress on this?

**Comment by chillyvee** [comment]
@adizere  Will try to make this a nice weekend project :)

**Comment by adizere** [comment]
> @adizere Will try to make this a nice weekend project :)

Hey, it's not urgent. I was asking because I was not able to test it. We spoke among the whole team also about the solution, and there's some concerns that the application state might be easily corruptible, or unsafe to transfer across machines. We are leaning towards a solution that reuses the snapshotting mechanism (instead of app.db).

**Comment by yihuang** [comment]
In my test, the current implementation do has some issues with handshake process, if it's tricky to hook into the handshake process, I think we can provide some low level tools instead, for example let user to download and store the consensus state at a arbitrary height using the light client protocol.

**Comment by cason** [comment]
@chillyvee Could you take a look on https://github.com/cometbft/cometbft/pull/801?

I think the proposal I am writing there, as ADR 104, is essentially the same as the one you are proposing.

**Comment by cason** [comment]
@yihuang State sync currently does what you are proposing, with the advantage that it makes sure that the application state matches CometBFT's state (i.e., block store and state store). It does not make much sense to me to bootstrap CometBFT's state to a given height, as you proposed in #803, without having an application at the same height.

In #801 we propose a mechanism to allow what you are proposing as the first bullet, i.e., to restore an application snapshot from the local file system.

**Comment by yihuang** [comment]
> It does not make much sense to me to bootstrap CometBFT's state to a given height, as you proposed in https://github.com/cometbft/cometbft/pull/803, without having an application at the same height.

the point is split up the state-sync process into multiple commands to maximize flexibility, there'll be a [separate command to restore the app state from a local snapshot](https://github.com/yihuang/cronos/commit/543fa95cff6d6f663eb7501ed17ca6399df71714#diff-4c62c6d6a679bf5409b5e60050c1ca281445cff3ef8b0c72a4375dcfcf7ce2e6), it could be implemented in cosmos-sdk

**Comment by yihuang** [comment]
BTW, the second bullet in your proposal can also be implemented in app side, as [demonstrated here](https://github.com/yihuang/cronos/commit/543fa95cff6d6f663eb7501ed17ca6399df71714#diff-4c62c6d6a679bf5409b5e60050c1ca281445cff3ef8b0c72a4375dcfcf7ce2e6)

**Comment by cason** [comment]
> the point is split up the state-sync process into multiple commands to maximize flexibility

Yes, I agree with this. Maybe not really CLI commands, but to make the steps more clearly separated on the code, although they are already kind of.

More specifically:

1. The [node startup](https://github.com/cometbft/cometbft/blob/ca4bb9ee5550d1745e9e3f8082459b8a8b389b17/node/setup.go#L472) we configure the Light client provider and call State Sync
2. State sync main [method](https://github.com/cometbft/cometbft/blob/ca4bb9ee5550d1745e9e3f8082459b8a8b389b17/statesync/syncer.go#L241) receives a potential snapshot, uses its height to start the light client, then use the light-client obtained `state` to verify the installed snapshot
3. Back to the node startup, the `state` and `commit` obtained by the light client are used to bootstrap the state store and the block store accordingly

**Comment by cason** [comment]
> I think i'm just trying to do the same thing without intrusive changes in the normal node start up logic

The ADR 104 proposal (#801) is not intrusive as well. It has the advantage of also considering the server side, namely, providing a command to allow "exporting" an application snapshot to a "file", that will be use to bootstrap another node.

**Comment by cason** [comment]
> BTW, the second bullet in your proposal can also be implemented in app side, as [demonstrated here](https://github.com/yihuang/cronos/commit/543fa95cff6d6f663eb7501ed17ca6399df71714#diff-96febaffa2123156f470328444e27e21fdd05798144f0658d3da9e28cc3b691cR13)

Yes, this is an alternative for the server side to "export" snapshots.

I am not sure that is the best one, as it requires _every_  snapshot to be saved to disk, even if the snapshots are not shipped out-of-band to fresh clients. Notice that snapshots are already saved somewhere by the application, and this approach with create some data duplication, with the disadvantage that the application knows how and when to prune or delete old snapshots, feature that would not trivially be transferred to this file system "snapshot store".

**Comment by yihuang** [comment]
> > BTW, the second bullet in your proposal can also be implemented in app side, as [demonstrated here](https://github.com/yihuang/cronos/commit/543fa95cff6d6f663eb7501ed17ca6399df71714#diff-96febaffa2123156f470328444e27e21fdd05798144f0658d3da9e28cc3b691cR13)
> 
> Yes, this is an alternative for the server side to "export" snapshots.
> 
> I am not sure that is the best one, as it requires _every_ snapshot to be saved to disk, even if the snapshots are not shipped out-of-band to fresh clients. Notice that snapshots are already saved somewhere by the application, and this approach will create some data duplication, with the disadvantage that the application knows how and when to prune or delete old snapshots, feature that would not be trivially transferred to this file system "snapshot store".

in cosmos-sdk, it'll directly restore the state from snapshot without saving them, it only store the snapshots created by itself later on, the new flag is to make it save the snapshot for the first restoration. cometbft do save the chunk to file in a temporary directory though, it's immediately removed after restoration finished.
the saved snapshot is to be used by another cli command to restore app state from it, then user use the `bootstrap-state` cli command to restore cometbft state, then a normal node start up.

**Comment by cason** [comment]
> in cosmos-sdk

We have to also consider users that do not adopt the SDK. Of course, once we have a more detailed proposal, we would be happy to discuss specifics with the SDK team.

**Comment by cason** [comment]
Moreover, I would really appreciate your feedback on #801. It is in draft mode, but you can comment or propose changes there.

**Comment by yihuang** [comment]
> Moreover, I would really appreciate your feedback on #801. It is in draft mode, but you can comment or propose changes there.

have left review comments, I think we need more description of the new design to decide if it's a better one.

**Comment by yihuang** [comment]
But I hope we can merge #803 to unblock app side solutions immediately, then we can take our time to find more ideal design, wdyt?

**Comment by adizere** [comment]
> > But I hope we can merge #803 to unblock app side solutions immediately, then we can take our time to find more ideal design, wdyt?
> 
> #803 is ported to sdk directly, it turns out we can hack tendermint state directly, so the whole local snapshot restoration procedure can be done in app side alone.

Excellent. Thanks @yihuang for the update!

I don't fully understand the SDK-side solution, but it seems to handle snapshot restoration plus also saving the snapshot (https://github.com/cosmos/cosmos-sdk/pull/16060). What exactly is left on Comet side to do?

**Comment by yihuang** [comment]
> > > But I hope we can merge #803 to unblock app side solutions immediately, then we can take our time to find more ideal design, wdyt?
> > 
> > 
> > #803 is ported to sdk directly, it turns out we can hack tendermint state directly, so the whole local snapshot restoration procedure can be done in app side alone.
> 
> Excellent. Thanks @yihuang for the update!
> 
> I don't fully understand the SDK-side solution, but it seems to handle snapshot restoration plus also saving the snapshot ([cosmos/cosmos-sdk#16060](https://github.com/cosmos/cosmos-sdk/pull/16060)). What exactly is left on Comet side to do?

maybe we can do [this command](https://github.com/cosmos/cosmos-sdk/pull/16061/files#diff-f946934088bd1a5b75846b7261731373d0b1f00ea507568ceb30fa8d624c6b4aR262) more elegantly in cometbft, after all, it use many internal APIs to do the job, those are maybe prone to change in the future.

**Comment by adizere** [comment]
> maybe we can do [this command](https://github.com/cosmos/cosmos-sdk/pull/16061/files#diff-f946934088bd1a5b75846b7261731373d0b1f00ea507568ceb30fa8d624c6b4aR262) more elegantly in cometbft, after all, it use many internal APIs to do the job, those are maybe prone to change in the future.

Got it. This is a nice to have, right? Correct me if I'm wrong, it seems like on Comet side there's nothing left to do.

**Comment by yihuang** [comment]
> > maybe we can do [this command](https://github.com/cosmos/cosmos-sdk/pull/16061/files#diff-f946934088bd1a5b75846b7261731373d0b1f00ea507568ceb30fa8d624c6b4aR262) more elegantly in cometbft, after all, it use many internal APIs to do the job, those are maybe prone to change in the future.
> 
> Got it. This is a nice to have, right? Correct me if I'm wrong, it seems like on Comet side there's nothing left to do (except for "nice to have" in the future).

yeah, nice to have.

**Comment by adizere** [comment]
> yeah, nice to have.

Thanks @yihuang. We'll track that work here: https://github.com/cometbft/cometbft/issues/884.

I'm closing the present PR: The original scope and problem we wanted to address here shifted. In the future, we might implement ADR 104 (#801) as a solution for non-SDK users that want support for state sync out of a local snapshot. For the case of SDK users, their needs seem to be addressed.
